src/bifrost/core.clj

```diff
@@ -41,86 +41,110 @@
-    {:enter
-     (fn [ctx]
-       (let [response-channel (async/chan 1 response-channel-xf)]
-         (async/>!! channel [response-channel ctx])
-         (assoc-in ctx [:response-channels response-channel-key] response-channel)))
-     :leave
-     (fn [ctx]
-       (if-let [response (async/alt!!
-                           (async/timeout timeout)
-                           timeout-response
-
-                           (get-in ctx [:response-channels response-channel-key])
-                           ([r] r))]
-         (merge ctx response)
-         ctx))})))
+ (def channel-closed-response
+  {:response {:status 500
+              :body "Channel was closed unexpectedly"}})
 
 (defn response-channel-key []
   (keyword (gensym "bifrost-response-channel-")))
 
+(defn interceptor*
+  "The main interceptor logic. Saves a channel on the context in the `:enter`
+  handler, and tries to take off of it in the `:leave` handler. Assocs a 504
+  response if `timeout` is reached before the channel has a result.
+
+  `->chan` is a function taking a context and returning an async channel that
+  will eventually contain a response.
+
+  `->response` is a function taking a (possibly nil) response that returns data
+  to merge into the context."
+  [->chan ->response timeout]
+  (let [response-channel-key (response-channel-key)]
+    (interceptor/interceptor
+     {:name ::interceptor
+      :enter
+      (fn [ctx]
+        (assoc-in ctx [:response-channels response-channel-key]
+                  (->chan ctx)))
+      :leave
+      (fn [ctx]
+        (let [response (async/alt!!
+                        (async/timeout timeout)
+                        timeout-response
+
+                        (get-in ctx [:response-channels response-channel-key])
+                        ([r]
+                         (->response r)))]
+          (merge ctx response)))})))
+
 (defn fn-interceptor
-  ([f timeout]
-   (let [response-channel-key (response-channel-key)]
-     (interceptor/interceptor
-      {:enter
-       (fn [ctx]
-         (let [response-channel (f (ctx->bifrost-request ctx))]
-           (assoc-in ctx [:response-channels response-channel-key]
-                     response-channel)))
-       :leave
-       (fn [ctx]
-         (if-let [response (async/alt!!
-                             (async/timeout timeout)
-                             timeout-response
-
-                             (get-in ctx [:response-channels response-channel-key])
-                             ([r]
-                              (when r
-                                (api-response->ctx r))))]
-           (merge ctx response)
-           ctx))}))))
+  [f timeout]
+  (interceptor* (comp f ctx->bifrost-request)
+                api-response->ctx
+                timeout))
+
+(defn- channel-adapter
+  "Takes an async `channel` and returns a function that can be used with
+  `interceptor*`."
+  [channel]
+  (let [request-ch (async/chan 1)]
+    (async/pipe request-ch channel)
```

Comment:
This is kind of weird, but the original code used `async/pipe` to effectively add a `(map ctx->bifrost-request)` xf, which I think has the side-effect of adding a 1-element buffer (which the tests relied on). So I just kept the dummy pipe here to keep the code simpler.